pkg/cloudprovider/types.go

```diff
@@ -80,14 +80,13 @@ type InstanceType interface {
 	// Overhead is the amount of resource overhead expected to be used by kubelet and any other system daemons outside
 	// of Kubernetes.
 	Overhead() v1.ResourceList
-	// Price is a metric that is used to optimize pod placement onto nodes as well as determining if replacing a node with
-	// a cheaper instance type is possible when performing consolidation.
-	Price() float64
 }
 
 // An Offering describes where an InstanceType is available to be used, with the expectation that its properties
 // may be tightly coupled (e.g. the availability of an instance type in some zone is scoped to a capacity type)
 type Offering struct {
 	CapacityType string
 	Zone         string
+	Price        float64
+	Available    bool
```

Comment:
Previously we used lack of existence as a signal for availability. Is there a reason this doesn't work anymore?

Reply:
The thought was that we could send back all the offerings that have ever existed and keep the pricing information in the offerings this way to keep it decoupled from the pricing provider. This was needed so that we could get a close price to use to calculate the savings on the node during consolidation.

In my mind there's a couple of options:
1. We keep track of the last price of the offering through this model, keeping availability in the offering and filtering out the offerings that are available when we needs to
2. Return a "close enough" price when we are trying to get the offering price in the consolidation loop (by getting either the exact offering or one within the same capacity type if that exact offering is no longer available (due to ICE or otherwise)), else return an error. This means we could leave the `Offerings()` call the same and we wouldn't have to change a ton of logic in the pricing provider.

Personally, I'm more in favor of the second since I think it keeps our `Offerings()` call much simpler but I know that @bwagner5 was thinking there may be other use-cases for having details around whether an offering _ever_ existed as part of the offerings logic.

Reply:
> This was needed so that we could get a close price to use to calculate the savings on the node during consolidation.

Ah -- I see. The problem is that we need to know the existing price of a node when performing a consolidation action. This actually sounds like a bug, since the price returned by the cloud provider can be different. 

What if we made price a well known label? This would enable us to consolidate based on an actual known value of the node, rather than guessing by trying to match with a result from Instance Types. IIUC from your statements above, this lets us remove the concept of availability from the API. 

This is extremely interesting from a selector perspective using the new GT/LT operators, since pods/provisioners could request capacity as long as it's under a certain price 😮.

> @bwagner5 was thinking there may be other use-cases for having details around whether an offering ever existed as part of the offerings logic.

I'm hesitant to build features/api surface before we know (or can at least guess) real use cases for them. There's some discussion on this in https://karpenter.sh/docs/contributing/design-guide/

Reply:
> What if we made price a well known label? 
Isn't the price in the fleet request?
Can't we access that one that is linked to the ec2 instance metadata?

Reply:
Yes, this would let us remove `.Availability` from the API. @bwagner5 and I discussed adding this label. It sounds like this is something we would have to add post-node creation since we won't know the instance type that will be created until after the node is actually joined and brought up. 

> This actually sounds like a bug, since the price returned by the cloud provider can be different.

I'm not quite getting how this would be a bug since the current price of the offering should be someone is currently getting charged (save special enterprise pricing). The issue I see with applying a node label is that pricing is variable and the pricing that is in the node label that is first used upon node creation may vary greatly over a long period of time (particularly for spot pricing).

Reply:
The price can also change per hour for Spot, so we'd need to reconcile the nodes with price.

Reply:
It could be an interesting example of the first provider specific reconciler.

Reply:
Resolved this discussion, the plan is to keep the `.Available` as a portion of the offerings and return all `Offerings` in the interface and filter out availability as a first-class helper function